gusto/time_discretisation.py

```diff
@@ -234,6 +234,128 @@ def apply(self, x_out, x_in):
         pass
 
 
+class ImplicitMultistage(TimeDiscretisation):
+    """
+    A class for implementing general diagonally implicit multistage (Runge-Kutta)
+    methods based on its Butcher tableau.
+
+    A Butcher tableau is formed in the following way for a s-th order
+    diagonally implicit scheme:
+
+    c_0 | a_00 a_01  .    a_0s
+    c_1 | a_10 a_11       a_1s
+     .  |   .   .    .    .
+     .  |   .   .    .    .
+    c_s | a_s0 a_s1  .    a_ss
+     -------------------------
+        |  b_1  b_2  ...  b_s
+
+    The gradient function has no time-dependence, so c elements are not needed.
+    A square 'butcher_matrix' is defined in each class that uses
+    the ImplicitMultiStage structure,
+
+    [a_00   0   .       0  ]
+    [a_10 a_11  .       0  ]
+    [  .    .   .       .  ]
+    [ b_0  b_1  .       b_s]
+
+    Unlike the explicit method, all upper diagonal a_ij elements are non-zero for implicit methods.
+
+    There are three steps to move from the current solution, y^n, to the new one, y^{n+1}
+
+    For an s stage method,
+      At iteration i (from 0 to s-1)
+        An intermediate location is computed as y_i = y^n + sum{over j less than or equal to i} (dt*a_ij*k_i)
+        Compute the gradient at the intermediate location, k_i = F(y_i)
+
+    At the last stage, compute the new solution by:
+    y^{n+1} = y^n + sum_{j from 0 to s} (dt*b_i*k_i)
+
+    """
+
+    def __init__(self, domain, field_name=None, solver_parameters=None, limiter=None, options=None, butcher_matrix=None):
```

Comment:
I think `butcher_matrix` needs to be a compulsory argument here (and so the second argument), since I think this scheme won't work unless we've specified the Butcher matrix. I'm afraid the same thing again applies to the `ExplicitMultistage` scheme...